### PR TITLE
Add new search arguments

### PR DESCRIPTION
**Pull request by fidgetingbits**
This pull request adds three additional search features:

- `-i`/`--interval` to specify the interval between each search result. This is
  useful if you're analyzing something like heap spray where you mostly just
  care if the memory you sprayed exists within some address range, and aren't
  interested in any of the subsequent search hits within that same range. So for instance
  if we're spraying a bunch of data and we want to make sure that every 1MB we hit some
  target address we can do:

```
pwndbg> search -i 0x100000 -t qword 0x0000000000010c00 [anon_97d82]
Searching for value: b'\x00\x0c\x01\x00\x00\x00\x00\x00'
[anon_97d82]    0x982e4d13 0x10c00
[anon_97d82]    0x98327dcf 0x10c00
[anon_97d82]    0x98400038 0x10c00
[anon_97d82]    0x98500038 0x10c00
[anon_97d82]    0x98600038 0x10c00
...
```

We can see that it only prints the first result from each 0x100000 block.

There's two annoyances with doing a search like this, which are addressed by the
other two features.

First, it can be exceptionally slow if the region of memory
are searching in is large (and doing remote debugging), and if you're only
interested in say the first ten results, you don't want to wait around to search 
through the entire region. The existing `trunc_out` feature doesn't help here
because it only truncates the output after a complete search is already completed.

Any example I'm using above the region is 512MB, and it takes a while to search remotely.

```
0x97d82000 0xb674c000 rw-p 1e9ca000      0 [anon_97d82]
```

This problem is addressed by this new option:

- `-l`/`--limit` to specify the maximum number of search results to return.
  This is useful if you're only interested in the first few results, and don't
  want to wait for the entire search to complete. For instance if we're
  spraying a bunch and we want to make sure that every 1MB we hit the target
  address we can do:

```
pwndbg> search -i 0x100000 -l 5 -t qword 0x0000000000010c00 [anon_97d82]
Searching for value: b'\x00\x0c\x01\x00\x00\x00\x00\x00'
[anon_97d82]    0x982e4d13 0x10c00
[anon_97d82]    0x98327dcf 0x10c00
[anon_97d82]    0x98400038 0x10c00
[anon_97d82]    0x98500038 0x10c00
[anon_97d82]    0x98600038 0x10c00
```

Second, some of the results above are not actually the data that were interested in,
for instance the two examples above happen to be the wanted value but at an
alignment that our spray will never actually hit, as in my example case I know
that what I spray will always be 8-byte aligned.

- `-a`/`--align` to specify the alignment of the search results. This is useful
  if you're only interested in results that are aligned to a certain value.
  For instance if we're spraying a bunch and we want to make sure that every
  1MB we hit the target address we can do:

```
pwndbg> search -i 0x100000 -a 0x8 -l 5 -t qword 0x0000000000010c00 [anon_97d82]
Searching for value: b'\x00\x0c\x01\x00\x00\x00\x00\x00'
[anon_97d82]    0x98387510 0x10c00
[anon_97d82]    0x98400038 0x10c00
[anon_97d82]    0x98500038 0x10c00
[anon_97d82]    0x98600038 0x10c00
[anon_97d82]    0x98700f78 0x10c00
```

You can see that earlier results that were at unwanted alignments no
longer show up.

I've included tests which pass on my machine.

As for the name of the commands, I suspect people might not like interval for
instance, I'm happy to change them to whatever you want.

Lastly the try except around the search_memory() call may be a bit questionable, and probably the format of the warning is wrong, so happy to change that if needed.

### CODE REVIEW

File: pwndbg/search.py [hunk 1]
```diff
@@ -58,7 +72,16 @@ def search(searchfor, mappings=None, start=None, end=None, executable=False, wri
...
 
-            start = i.search_memory(start, length, searchfor)
+            try:
+                start = i.search_memory(start, length, searchfor)
+            except gdb.error:
+                # While remote debugging on an embedded device and searching
+                # through a large memory region (~512mb), gdb may return an error similar
+                # to `error: Invalid hex digit 116`, even though the search
+                # itself is ok. It seems to have to do with a timeout.
+                print("WARN: gdb.search_memory failed, stopping early...")
```
Comment (disconnect3d) [inline pwndbg/search.py]: Ideally this should be a separate PR/commit if possible, but I'm okay leaving this here.

On the other hand, should we maybe check the gdb.error message string or print it to the user as well?

If we think this may be a timeout, we could say it to the user as well that this may happen for embedded debugging

Comment (fidgetingbits) [reply pwndbg/search.py]: I'll break it out into a separate PR then. 

Confusingly the gdb error is `error: Invalid hex digit 116`, but from a bit of digging it seems like this sometimes comes from gdb getting confused during timeouts. The unfortunate thing is that error isn't exactly useful to relay back to the user, but I guess we may as well reported back.

**Comment by disconnect3d** [comment]
I hope to review this on this weekend

File: pwndbg/commands/search.py [hunk 2]
```diff
@@ -132,7 +144,20 @@ def print_search_hit(address) -> None:
...
+    executable,
+    writable,
+    interval,
+    limit,
+    aligned,
+    value,
+    mapping_name,
+    save,
+    next,
+    trunc_out,
```
Comment (disconnect3d) [inline pwndbg/commands/search.py:171]: Just to be sure: the trunc_out still makes sense if we consider saving of results, right?

Comment (fidgetingbits) [reply pwndbg/commands/search.py:171]: Ya that was my thinking and why I left them separate. I suppose it may be worth documenting that in the usage somewhere to make the distinction clear?

Comment (disconnect3d) [reply pwndbg/commands/search.py:171]: Yeah, ideally we should probably also add example usages in the docstring too

Comment (disconnect3d) [reply pwndbg/commands/search.py:171]: (for that https://github.com/pwndbg/pwndbg/blob/dev/pwndbg/commands/vmmap.py#L46 would need to be used so that we can use newlines in the docstring heh)

**Comment by disconnect3d** [comment]
Btw this looks good to me but it needs to be rebased now to be merged

File: pwndbg/commands/search.py [hunk 3]
```diff
@@ -105,6 +104,19 @@ def print_search_hit(address) -> None:
     "-e", "--executable", action="store_true", help="Search executable segments only"
 )
 parser.add_argument("-w", "--writable", action="store_true", help="Search writable segments only")
+parser.add_argument(
+    "-i",
+    "--interval",
```
Comment (disconnect3d) [inline pwndbg/commands/search.py]: Shall we call it `-d --distance`? Or any other ideas?

cc: @gsingh93 @Arusekk @lebr0nli for opinion

Comment (fidgetingbits) [reply pwndbg/commands/search.py]: fwiw to me interval makes sense (which is why I named it that way), as a search interval implies a repetition with a fixed value. I just am not sure it would be obvious to others, especially none-native english speakers. But --distance doesn't imply repetition to me so I think might be even more confusing.

Comment (disconnect3d) [reply pwndbg/commands/search.py]: Let's do `-s --step`

File: pwndbg/commands/search.py (hunk 2 above)
Comment (fidgetingbits) [reply pwndbg/commands/search.py:171]: I added a couple small comments in the same structure as vmmap just commenting on the difference, but I think probably to add the full documentation for the commands with examples depending on how you want to do it would be a different pull request so not going todo it for now.

**Comment by fidgetingbits** [comment]
I added a couple more basic test cases, because I randomly ran into a separate bug that didn't actually end up being related to this (#808 seemed to trigger on a new gdb client version but with a older gdb server, but I'll file a different bug).

I also added a separate check to make sure that there's actually some map found to search, or error otherwise. I can remove that if you want it to be a separate PR

Otherwise I just addressed your comments about switching interval to step, and re-based, so I think it should be okay to merge now

**Comment by fidgetingbits** [comment]
Ah I seemed to mess up the rebase, so its showing a bunch of other changed files now... will have to fix that
